### PR TITLE
IGNITE-22386 Fix revision serialization where it is required to be properly comparable

### PR DESCRIPTION
**Pull request by ibessonov**
https://issues.apache.org/jira/browse/IGNITE-22386

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/util/ByteUtils.java [hunk 1]
```diff
@@ -64,6 +65,17 @@ public static long bytesToLong(byte[] bytes) {
...
     }
 
+
+    /**
+     * Constructs {@code long} from byte array created with {@link #longToComparableBytes(long)} with offset equal to 0.
+     *
+     * @param bytes Array of bytes.
+     * @return Long value.
+     */
+    public static long comparableBytesToLong(byte[] bytes) {
```
Comment (sanpwc) [inline modules/core/src/main/java/org/apache/ignite/internal/util/ByteUtils.java:74]: What's the penalty of such serialisation/deserialisation in comparison to `longToBytes` that forces you to have both? I mean that there's a chance that somebody will miss the fact the there's an extra comparable version of the method and will use common one where he actually need comparison.

File: modules/core/src/main/java/org/apache/ignite/internal/util/ByteUtils.java [hunk 2]
```diff
@@ -74,6 +86,17 @@ public static byte[] longToBytes(long l) {
         return putLongToBytes(l, new byte[8], 0);
     }
 
+    /**
```
Comment (sanpwc) [inline modules/core/src/main/java/org/apache/ignite/internal/util/ByteUtils.java:88]: What about ints?

File: modules/core/src/main/java/org/apache/ignite/internal/util/ByteUtils.java [hunk 3]
```diff
@@ -74,6 +85,17 @@ public static byte[] longToBytes(long l) {
...
+     * Converts a primitive {@code long} value to a byte array than can be compared with {@link ByteBuffer#compareTo(ByteBuffer)} with the
```
Comment (rpuch) [inline modules/core/src/main/java/org/apache/ignite/internal/util/ByteUtils.java:89]:
```suggestion
     * Converts a primitive {@code long} value to a byte array that can be compared with {@link ByteBuffer#compareTo(ByteBuffer)} with the
```

File: modules/core/src/main/java/org/apache/ignite/internal/util/ByteUtils.java [hunk 4]
```diff
@@ -74,6 +85,17 @@ public static byte[] longToBytes(long l) {
...
+     * same order as the original {@code long}.
+     *
+     * @param l Long value.
+     * @return Array of bytes.
+     */
+    public static byte[] longToComparableBytes(long l) {
```
Comment (rpuch) [inline modules/core/src/main/java/org/apache/ignite/internal/util/ByteUtils.java:95]: `Comparable` in the name is a bit confusing, it's not clear, comparable to what and why the alternative is not comparable (as it is comparable, but *differently*).

How about `longToBytesKeepingComparability()`? Or `longToBytesKeepingOrder()`

File: modules/core/src/main/java/org/apache/ignite/internal/util/ByteUtils.java (hunk 1 above)
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/internal/util/ByteUtils.java:74]: No penalty at all. If developers won't try deserializing data by hand and will always use these methods, then there would be no difference at all

File: modules/core/src/main/java/org/apache/ignite/internal/util/ByteUtils.java (hunk 2 above)
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/internal/util/ByteUtils.java:88]: No one needs it right now

File: modules/core/src/test/java/org/apache/ignite/internal/util/ByteUtilsTest.java [hunk 5]
```diff
@@ -43,4 +48,19 @@ void testStringFromBytes() {
         assertEquals("", stringFromBytes(stringToBytes("")));
         assertEquals("abc", stringFromBytes(stringToBytes("abc")));
     }
+
+    @Test
+    void testComparableLong() {
```
Comment (rpuch) [inline modules/core/src/test/java/org/apache/ignite/internal/util/ByteUtilsTest.java:53]: The method is pretty subtle, so we should probably have more (separate) tests. I would suggest to have the following:

1. A test that makes sure that forward-then-backward conversion yields the same result we had in the beginning (this is checked by the first line of the test)
2. For each of 8 bytes, a test that makes sure that, if just Nth byte differs, and in both cases the value of the byte has the same sign, the order is as expected
3. For each of 8 bytes, a test that makes sure that, if just Nth byte differs, and the values of the byte has different signs, the order is as expected

Parameterization could be used for 2 and 3.

File: modules/core/src/main/java/org/apache/ignite/internal/util/ByteUtils.java (hunk 1 above)
Comment (sanpwc) [reply modules/core/src/main/java/org/apache/ignite/internal/util/ByteUtils.java:74]: So, my question is actually following. Should we include xoring into common longToBytes, intToBytes, etc in order to preserve comparison in lex-order and remove comparable<> counterparts that you've added?

File: modules/core/src/main/java/org/apache/ignite/internal/util/ByteUtils.java (hunk 2 above)
Comment (sanpwc) [reply modules/core/src/main/java/org/apache/ignite/internal/util/ByteUtils.java:88]: So, someone else will have some fun while debugging similar issue in a year with ints, right? ;)

File: modules/core/src/test/java/org/apache/ignite/internal/util/ByteUtilsTest.java (hunk 5 above)
Comment (ibessonov) [reply modules/core/src/test/java/org/apache/ignite/internal/util/ByteUtilsTest.java:53]: Isn't that an overkill? It's really hard to implement this method wrongly

File: modules/core/src/main/java/org/apache/ignite/internal/util/ByteUtils.java (hunk 1 above)
Comment (rpuch) [reply modules/core/src/main/java/org/apache/ignite/internal/util/ByteUtils.java:74]: What if someone wants to get a BE representation specifically?

File: modules/core/src/main/java/org/apache/ignite/internal/util/ByteUtils.java (hunk 2 above)
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/internal/util/ByteUtils.java:88]: Maybe

File: modules/core/src/test/java/org/apache/ignite/internal/util/ByteUtilsTest.java (hunk 5 above)
Comment (ibessonov) [reply modules/core/src/test/java/org/apache/ignite/internal/util/ByteUtilsTest.java:53]: I also have check where different bytes differ (1 vs 256), it validates that we use BE

File: modules/core/src/main/java/org/apache/ignite/internal/util/ByteUtils.java (hunk 1 above)
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/internal/util/ByteUtils.java:74]: We already have instances where these methods are used exclusively as a conversion to BE. What do we do with them?

Comment (sanpwc) [reply modules/core/src/main/java/org/apache/ignite/internal/util/ByteUtils.java]: Discussed in private. Up to you.

File: modules/core/src/main/java/org/apache/ignite/internal/util/ByteUtils.java (hunk 2 above)
Comment (sanpwc) [reply modules/core/src/main/java/org/apache/ignite/internal/util/ByteUtils.java:88]: Well, I strongly suggest to add comparable counterparts to int.

File: modules/core/src/test/java/org/apache/ignite/internal/util/ByteUtilsTest.java (hunk 5 above)
Comment (rpuch) [reply modules/core/src/test/java/org/apache/ignite/internal/util/ByteUtilsTest.java]: There is a constant with 8 and 0 intermingled. It's easy to mess them up in this pattern, and not so easy to spot...

**Review by rpuch — APPROVED** [review]

**Review by sanpwc — APPROVED** [review]

**Pull request merged**